system/Test/DOMParser.php

```diff
@@ -61,8 +61,9 @@ public function getBody(): string
      */
     public function withString(string $content)
     {
-        // converts all special characters to utf-8
-        $content = mb_convert_encoding($content, 'HTML-ENTITIES', 'UTF-8');
```

Comment:
@MGatner To begin with, why do we need this convertion?
The comment is wrong. It never converts to utf-8.

```php
$str_utf8 = '<s>&lt;div&gt;あいう&lt;/div&gt;</s>';
$str_sjis = mb_convert_encoding($str_utf8, 'SJIS-win', 'utf8');

$encoded = mb_convert_encoding($str_utf8, 'HTML-ENTITIES', 'UTF-8');
echo $encoded . PHP_EOL;

$encoded = mb_convert_encoding($str_sjis, 'HTML-ENTITIES', 'UTF-8');
echo $encoded . PHP_EOL;
```
Results:
```
<s>&lt;div&gt;&#12354;&#12356;&#12358;&lt;/div&gt;</s>
<s>&lt;div&gt;&#4294967295;&#4294967295;&#4294967295;&#4294967295;&#4294967295;&#4294967295;&lt;/div&gt;</s>
```

Reply:
If I remove the line, tests in DOMParserTest fail.

Reply:
I have no idea why this is here. Seems very strange to be part of our test DOM parser.

Reply:
It seems DOMDocument does not process Japanese correctly.

```php
        $html = 'こんにちは、世界';

        $dom = new DOMDocument('1.0', 'utf-8');
        $dom->loadHTML($html);
        dd($dom->saveHTML($dom->documentElement));
```
![Screenshot 2022-10-24 17 48 26](https://user-images.githubusercontent.com/87955/197486495-6645e326-19df-4498-84e2-f24197fca39d.png)

Reply:
This works.

```php
        $html = 'こんにちは、世界';

        $dom = new DOMDocument('1.0', 'utf-8');
        $html = mb_encode_numericentity($html, [0x80, 0x10FFFF, 0, 0x1FFFFF], 'UTF-8');
        $dom->loadHTML($html);
        var_dump($dom->saveHTML($dom->documentElement));
```
```
string(57) "<html><body><p>こんにちは、世界</p></body></html>"
```

Reply:
Maybe non-ASCII characters are not handled correctly, and we need to convert all of them to numeric character references.

Reply:
Have you tried setting the object encoding? https://stackoverflow.com/questions/6573258/domdocument-and-special-characters#6620546

Reply:
The `encoding` is specified in the second parameter of the constructor.

```php
$dom = new DOMDocument('1.0', 'utf-8');
```
And the stackoverflow sample code uses `utf8_decode()` that is also deprecated in 8.2.

Reply:
> DOMDocument::loadHTML will treat your string as being in ISO-8859-1 (the HTTP/1.1 default character set) unless you tell it otherwise. 
https://stackoverflow.com/questions/8218230/php-domdocument-loadhtml-not-encoding-utf-8-correctly/8218649#8218649

Reply:
I think `mb_encode_numericentity($html, [0x80, 0x10FFFF, 0, 0x1FFFFF], 'UTF-8')` is the best for this case.
Prepending `<?xml encoding="utf-8" ?>` changes the output.
```diff
--- Expected
+++ Actual
@@ @@
 '<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
-<html><body><div><h1>Hello</h1></div></body></html>
+<?xml encoding="utf-8" ?><html><body><div><h1>Hello</h1></div></body></html>
 '
```

Reply:
We could always pretend the XML tag and then cut it off again, but let's go with your solution since it seems cleaner.